File: decorator/c-sharp/DataSourceDecorator.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace Decorator
{
    // UML diyagramındaki Component (IDataSource) arayüzünü uygular.
    // Constructor sayesinde özellikleri değiştirilmesi istenen nesnenin referansı tutulur.
    // Bu sınıftan türeyen sınıfların uygulamasını, değiştirilmesini istediğimiz metotları abstract anahtar kelimesi ile işaretledik.
    // UML diyagramındaki BaseDecorator yapısına denk gelmektedir.
    abstract class BaseDataSourceDecorator : IDataSource
    {
        protected IDataSource dataSource;

        public BaseDataSourceDecorator(IDataSource dataSource)
        {
            this.dataSource = dataSource;
        }

        public abstract string GetFileName();

        public abstract void ReadData();

        public abstract void WriteData(object data);
    }
}

File: decorator/typescript/data-source-decorator.ts
import {IDataSource} from "./data-source.interface";

// UML diyagramındaki Component (IDataSource) arayüzünü uygular.
// Constructor sayesinde özellikleri değiştirilmesi istenen nesnenin referansı tutulur.
// Bu sınıftan türeyen sınıfların uygulamasını, değiştirilmesini istediğimiz metotları abstract anahtar kelimesi ile işaretledik.
// UML diyagramındaki BaseDecorator yapısına denk gelmektedir.
export abstract class BaseDataSourceDecorator implements IDataSource {
  protected constructor(protected readonly dataSource: IDataSource) {}

  abstract getFileName(): string;

  abstract readData(): void;

  abstract writeData(data: object): void
}
